### PR TITLE
[Messenger][Profiler] Trace middleware execution

### PR DESCRIPTION
**Pull request by ogizanagi**
| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | yes 
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | part of #27262   
| License       | MIT
| Doc PR        | N/A

This is a start for #27262 with:
- traceable Messenger middlewares
- ~~a dedicated category for http kernel controller args resolvers~~ => See #28387

screenshot 2018-05-20 a 12 23 55

Messenger middleware are traced, with bus info (if not shared accros buses):

screenshot 2018-05-20 a 12 28 15

Another possibility is to use the middleware id instead of the class (with or without extra bus info?):

screenshot 2018-05-20 a 12 32 24

(_of course, collected times are faked here using `usleep` in the traceable middleware_)

### CODE REVIEW

**Review by sroze — CHANGES_REQUESTED** [review]

**Comment by ogizanagi** [comment]
@sroze : Thanks for the review. I'll try to update this soon 👍

**Comment by ogizanagi** [comment]
PR updated. ("soon" wasn't really reliable here, sorry ^^')

File: src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php [hunk 1]
```diff
@@ -306,6 +310,20 @@ private function registerBusMiddleware(ContainerBuilder $container, string $busI
                 throw new RuntimeException(sprintf('Invalid middleware factory "%s": a middleware factory must be an abstract definition.', $id));
             }
 
+            if ($debug) {
+                $container->register($debugMiddlewareId = "debug.traced.$messengerMiddlewareId", TraceableMiddleware::class)
```
Comment (stof) [inline src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php]: I would use `messenger.debug.traced.` as prefix, to ensure that we don't get conflicts with other components wanting to do debug things in the future.

Comment (stof) [reply src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php]: and we might want to make it hidden

File: src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php [hunk 2]
```diff
@@ -306,6 +310,20 @@ private function registerBusMiddleware(ContainerBuilder $container, string $busI
...
+                    // Decorates with a high priority so it's applied the earliest:
+                    ->setDecoratedService($messengerMiddlewareId, null, 100)
+                    ->setArguments(array(
+                        new Reference("$debugMiddlewareId.inner"),
```
Comment (stof) [inline src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php]: use concatenation rather than interpolation, according to our existing codebase

File: src/Symfony/Component/Messenger/Envelope.php [hunk 3]
```diff
@@ -86,4 +86,15 @@ public function getMessage()
...
         return $this->message;
     }
+
+    /**
+     * @param object $target
+     *
+     * @return Envelope|object The original message or the envelope if the target supports it
+     *                         (i.e implements {@link EnvelopeAwareInterface}).
+     */
+    public function getMessageFor($target)
```
Comment (stof) [inline src/Symfony/Component/Messenger/Envelope.php]: IMO, this should rather be a private method in TraceableMiddleware instead of exposing this method here publicly

File: src/Symfony/Component/Messenger/Middleware/Enhancers/TraceableMiddleware.php [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        $this->busName = $busName;
+        $this->eventCategory = $eventCategory;
+    }
+
+    /**
+     * @param Envelope $message
+     */
+    public function handle($message, callable $next)
+    {
+        $eventName = \get_class($this->inner);
```
Comment (stof) [inline src/Symfony/Component/Messenger/Middleware/Enhancers/TraceableMiddleware.php]: this should add support for anonymous classes, as done in #28218

File: src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php [hunk 5]
```diff
@@ -306,6 +310,20 @@ private function registerBusMiddleware(ContainerBuilder $container, string $busI
...
+                        new Reference($this->debugStopwatch),
+                        // If definition isn't abstract, we can't be sure the service instance
+                        // is only used in one bus at the same time, so we don't inject the bus name in this case:
+                        $isAbstract ? $busId : null,
```
Comment (stof) [inline src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php]: the implementation seems to do exactly the opposite of the comment.

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/time.html.twig [hunk 6]
```diff
@@ -9,6 +9,8 @@
         'event_listener':         '#00B8F5',
         'template':               '#66CC00',
         'doctrine':               '#FF6633',
+        'controller.argument_value_resolver': '#8c5de6',
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/time.html.twig]: that one should belong to a different PR IMO, as it is not about Messenger

File: src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php (hunk 5 above)
Comment (ogizanagi) [reply src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php]: > If definition isn't abstract [...] we don't inject the bus name

Feels like it is to me. But perhaps can be reworded.

Comment (stof) [reply src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php]: but here, you do inject it *only* when it is abstract

**Comment by fabpot** [comment]
@ogizanagi Do you have time to finish this PR?

**Comment by ogizanagi** [comment]
@fabpot : Done. I've also extracted the first commit in another PR (#28387).

Status: Needs review

File: src/Symfony/Component/Messenger/Middleware/Enhancers/TraceableMiddleware.php [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+use Symfony\Component\Messenger\EnvelopeAwareInterface;
+use Symfony\Component\Messenger\Middleware\MiddlewareInterface;
+use Symfony\Component\Stopwatch\Stopwatch;
+
+/**
+ * Collects some data about a middleware.
+ *
+ * @author Maxime Steinhausser <maxime.steinhausser@gmail.com>
+ *
+ * @experimental in 4.1
```
Comment (chalasr) [inline src/Symfony/Component/Messenger/Middleware/Enhancers/TraceableMiddleware.php]: can be removed I guess

**Review by chalasr — APPROVED** [review]
Fine using FQCNs to me, should be easy to reconsider if needed.

**Review by sroze — APPROVED** [review]
(rebase needed though)
